lib/Service/ZipService.php

```diff
@@ -26,37 +26,70 @@
 
 namespace OCA\FilesZip\Service;
 
+use Exception;
+use OC\User\NoUserException;
+use OCA\FilesZip\BackgroundJob\ZipJob;
+use OCA\FilesZip\Exceptions\TargetAlreadyExists;
+use OCP\BackgroundJob\IJobList;
 use OCP\Files\File;
 use OCP\Files\Folder;
 use OCP\Files\IRootFolder;
 use OCP\Files\Node;
+use OCP\Files\NotFoundException;
+use OCP\Files\NotPermittedException;
+use OCP\IUserSession;
+use OCP\Lock\LockedException;
 use ZipStreamer\ZipStreamer;
 
 class ZipService {
 
 	/** @var IRootFolder */
 	private $rootFolder;
-
-	public function __construct(IRootFolder $rootFolder) {
+	/** @var NotificationService */
+	private $notificationService;
+	/** @var IUserSession */
+	private $userSession;
+	/** @var IJobList */
+	private $jobList;
+
+	public function __construct(IRootFolder $rootFolder, NotificationService $notificationService, IUserSession $userSession, IJobList $jobList) {
 		$this->rootFolder = $rootFolder;
+		$this->notificationService = $notificationService;
+		$this->userSession = $userSession;
+		$this->jobList = $jobList;
+	}
+
+	public function createZipJob(array $fileIds, $target): void {
+		$user = $this->userSession->getUser();
+		if ($user === null) {
+			throw new Exception('No user session available');
+		}
+		$this->jobList->add(ZipJob::class, [
+			'uid' => $user->getUID(),
+			'fileIds' => $fileIds,
+			'target' => $target,
+		]);
+		$this->notificationService->sendNotificationOnPending($user->getUID(), $target);
 	}
 
 	/**
-	 * @param string $uid
-	 * @param int[] $fileIds
-	 * @param string $target
-	 *
-	 * @throws \OCP\Files\NotPermittedException
-	 * @throws \OCP\Lock\LockedException
-	 * @throws \OC\User\NoUserException
+	 * @throws NotPermittedException
+	 * @throws NoUserException
+	 * @throws TargetAlreadyExists
+	 * @throws LockedException
 	 */
-	public function zip(string $uid, array $fileIds, string $target) {
+	public function zip(string $uid, array $fileIds, string $target): File {
 		$userFolder = $this->rootFolder->getUserFolder($uid);
 
-		// Todo obtain proper target node
-		// todo; verify node doesn't exist yet
-		$taregetNode = $userFolder->newFile('target.zip');
-		$outStream = $taregetNode->fopen('w');
+		try {
+			$userFolder->get($target);
+			throw new TargetAlreadyExists();
+		} catch (NotFoundException $e) {
+			// Expected behavior that the file does not exist yet
+		}
+
+		$targetNode = $userFolder->newFile($target);
```

Comment:
Will the zip file always get created at the users root?

Reply:
No, the target is specifying the full path which is sent over.

Reply:
See https://github.com/nextcloud/files_zip/pull/22/files#diff-27653c212e1cfe533e4eb2f7d0d3f89604c9de48a09583b4cbbbcbd08a07da79R14-R26 for how it is built :)

Reply:
Sounds good! Putting it at the users root folder would otherwise caused issues when the root storage is read only. Just wanted to raise awareness of this :)